src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+ *    limitations under the License.
+ */
+#include <app/codegen-interaction-model/CodegenDataModel.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/endpoint-config-api.h>
+#include <lib/core/DataModelTypes.h>
+#include <optional>
+
+namespace chip {
+namespace app {
+namespace {
+
+/// Checks if the specified ember cluster mask corresponds to a valid
+/// server cluster.
+bool IsServerMask(EmberAfClusterMask mask)
+{
+    return (mask == 0) || ((mask & CLUSTER_MASK_SERVER) != 0);
+}
+
+/// Load the cluster information into the specified destination
+void LoadClusterInfo(const ConcreteClusterPath & path, const EmberAfCluster & cluster, InteractionModel::ClusterInfo * info)
+{
+    chip::DataVersion * versionPtr = emberAfDataVersionStorage(path);
+    if (versionPtr != nullptr)
+    {
+        info->dataVersion = *versionPtr;
+    }
+    else
+    {
+        ChipLogError(AppServer, "Failed to get data version for %d/" ChipLogFormatMEI, static_cast<int>(path.mEndpointId),
+                     ChipLogValueMEI(cluster.clusterId));
+        info->dataVersion = 0;
+    }
+
+    // TODO: set entry flags:
+    //   info->flags.Set(ClusterQualityFlags::kDiagnosticsData)
+}
+
+/// Converts a EmberAfCluster into a ClusterEntry
+InteractionModel::ClusterEntry ClusterEntryFrom(EndpointId endpointId, const EmberAfCluster & cluster)
+{
+    InteractionModel::ClusterEntry entry;
+
+    entry.path = ConcreteClusterPath(endpointId, cluster.clusterId);
+    LoadClusterInfo(entry.path, cluster, &entry.info);
+
+    return entry;
+}
+
+/// Finds the first server cluster entry for the given endpoint data starting at [start_index]
+///
+/// Returns an invalid entry if no more server clusters are found
+InteractionModel::ClusterEntry FirstServerClusterEntry(EndpointId endpointId, const EmberAfEndpointType * endpoint,
+                                                       unsigned start_index, unsigned & found_index)
+{
+    for (unsigned cluster_idx = start_index; cluster_idx < endpoint->clusterCount; cluster_idx++)
+    {
+        const EmberAfCluster & cluster = endpoint->cluster[cluster_idx];
+        if (!IsServerMask(cluster.mask & CLUSTER_MASK_SERVER))
+        {
+            continue;
+        }
+
+        found_index = cluster_idx;
+        return ClusterEntryFrom(endpointId, cluster);
+    }
+
+    return InteractionModel::ClusterEntry::Invalid();
+}
+
+/// Load the cluster information into the specified destination
+void LoadAttributeInfo(const ConcreteAttributePath & path, const EmberAfAttributeMetadata & attribute,
+                       InteractionModel::AttributeInfo * info)
+{
+    if (attribute.attributeType == ZCL_ARRAY_ATTRIBUTE_TYPE)
+    {
+        info->flags.Set(InteractionModel::AttributeQualityFlags::kListAttribute);
+    }
+
+    // TODO: Set additional flags:
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kChangesOmitted)
```

Comment:
Are these flags set currently? What effect does leaving this as a TODO have on the way the code currently operates?

Reply:
I could not find a place where these two attributes are set or used.

Reply:
I currently believe these are not used at all. I think they should be however we have no metadata on it currently. We have to determine what to do here. For now just noting this should be a thing even though it does not seem implemented in the code.